pkg/cloud/gcp/actuators/util/gcp_tags_labels.go

```diff
@@ -119,3 +130,100 @@ func GetLabelsList(userLabelsAllowed bool, client controllerclient.Client, machi
 
 	return labels, nil
 }
+
+// getInfraResourceTagsList returns the user-defined tags present in the
+// status sub-resource of Infrastructure.
+func getInfraResourceTagsList(platformStatus *configv1.PlatformStatus) []machinev1.ResourceManagerTag {
+	if platformStatus != nil && platformStatus.GCP != nil && platformStatus.GCP.ResourceTags != nil {
+		tags := make([]machinev1.ResourceManagerTag, len(platformStatus.GCP.ResourceTags))
+		for i, tag := range platformStatus.GCP.ResourceTags {
+			tags[i] = machinev1.ResourceManagerTag{
+				ParentID: tag.ParentID,
+				Key:      tag.Key,
+				Value:    tag.Value,
+			}
+		}
+		return tags
+	}
+	return nil
+}
+
+// getTagValuesNames returns the list of tags in Compute APIs required format, which
+// is a map containing keys of the form Key(`tagKeys/{tag_key_id}`) and values in the
+// form (`tagValues/{tag_value_id}`).
+func getTagValuesNames(ctx context.Context, tagService tagservice.TagService, tagList []machinev1.ResourceManagerTag) (map[string]string, error) {
+	tagValueList := make(map[string]string, len(tagList))
+	for _, tag := range tagList {
+		name := fmt.Sprintf("%s/%s/%s", tag.ParentID, tag.Key, tag.Value)
+		value, err := tagService.GetNamespacedName(ctx, name)
+		if err != nil {
+			var gErr *apierror.APIError
+			// google API returns StatusForbidden or StatusNotFound when the tag
+			// does not exist, since it could be because of permission issues
+			// or genuinely tag does not exist.
+			if errors.As(err, &gErr) && (gErr.HTTPCode() == http.StatusNotFound ||
+				gErr.HTTPCode() == http.StatusForbidden) {
+				klog.Errorf("does not have permission to access %s tag or tag does not exist", name)
+				continue
+			}
+			return nil, err
```

Comment:
We intentionally fail if fetching a single tag fails? Is that a fatal error here? I guess this is only for temporary resource fetching errors right?

Can we wrap this with some context please?

Reply:
We should be able to fetch details of every configured tag, since its required to be passed during Compute resource creation. Other option I think would be to maintain state in status.Condition for tag specific errors and retry adding tags which failed here after resource is created. 

I removed ignoring the permission denied or non-existent tag errors, I think even those should be left to user to rectify, correct?

User configured tags contain parentID, tagKey and tagValue short names, which will be used for deriving tag's namespaced name of the form `parentID/tagKey/tagValue` and is used for fetching tag's key and value names of the form `tagKeys/<UID>`, `tagValues/<UID>` as expected by the create compute API.